test/unit/org/apache/cassandra/repair/RepairJobTest.java

```diff
@@ -0,0 +1,569 @@
+                                                         treeResponse(addr2, range1, "two",   range2, "two",   range3, "two"),
+                                                         treeResponse(addr3, range1, "three", range2, "three", range3, "three"));
+
+        Map<SyncNodePair, SyncTask> tasks = toMap(RepairJob.createOptimisedSyncingSyncTasks(desc,
+                                                                                            treeResponses,
+                                                                                            addr1, // local
+                                                                                            noTransient(),
+                                                                                            addr -> "DC1",
+                                                                                            false,
+                                                                                            PreviewKind.ALL));
+
+        for (SyncNodePair pair : new SyncNodePair[]{ pair(addr1, addr2),
+                                                     pair(addr1, addr3),
+                                                     pair(addr2, addr1),
+                                                     pair(addr2, addr3),
+                                                     pair(addr3, addr1),
+                                                     pair(addr3, addr2) })
+        {
+            assertEquals(Arrays.asList(range1, range2, range3), tasks.get(pair).rangesToSync);
+        }
+    }
+
+    @Test
+    public void testOptimizedCreateStandardSyncTasks()
+    {
+        List<TreeResponse> treeResponses = Arrays.asList(treeResponse(addr1, range1, "one",   range2, "one"),
+                                                         treeResponse(addr2, range1, "one",   range2, "two"),
+                                                         treeResponse(addr3, range1, "three", range2, "two"));
+
+        Map<SyncNodePair, SyncTask> tasks = toMap(RepairJob.createOptimisedSyncingSyncTasks(desc,
+                                                                                            treeResponses,
+                                                                                            addr4, // local
+                                                                                            noTransient(),
+                                                                                            addr -> "DC1",
+                                                                                            false,
+                                                                                            PreviewKind.ALL));
+
+        assertEquals(Arrays.asList(range1), tasks.get(pair(addr1, addr3)).rangesToSync);
+        assertEquals(Arrays.asList(range2), tasks.get(pair(addr1, addr2)).rangesToSync);
```

Comment:
this assert depends on the iteration order in https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/repair/asymmetric/ReduceHelper.java#L55 - when we have an option to stream between two hosts, we always pick the one with the fewest outgoing streams, so if by chance we have added a few outgoing streams to `addr2` before this, we might sync `range2` from `addr3` instead

we should probably assert that `addr1` gets `range2` from either `addr2` or `addr3` (but not both) instead

Reply:
This invariant does not hold in the current code: `range2` _is_ streamed from both `addr1` and `addr2`, assert below checks for that. This doesn't look right, but I'd have to dig deeper to understand why.